Support remote debugging

pvid opened the pull request:
A first draft of adding support of remote debugging in bloop.

It seems that a new value for `DebugSessionParamsDataKind ` should be addend to BSP. I'd love your input if this is the way to go and what else needs to be changed or updated.

I have added a "happy path" test so far and I have also done some manual tests using a version of `metals` with necessary changes

![test-workspace-1601236890032](https://user-images.githubusercontent.com/32688476/94374808-08e08080-010f-11eb-8bf2-5d19adfb1f1f.gif)

tgodzik commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -586,6 +586,9 @@ final class BloopBspServices(
         case bsp.DebugSessionParamsDataKind.ScalaTestSuites =>
           convert[List[String]](filters => DebuggeeRunner.forTestSuite(projects, filters, state))
+        // TODO incorporate into BSP, maybe?
+        case "scala-attach-remote" =>
```
I think that's a good call.

marek1840 commented on frontend/src/main/scala/bloop/dap/DebugSession.scala:
```diff
@@ -178,6 +178,9 @@ final class DebugSession(
         response.command = Command.LAUNCH.getName
         attachedPromise.success(())
         super.sendResponse(response)
+      case "attach" =>
```
Please, add comment that it is a response to an actual "attach" request and not our synthetic one

pvid replied:
I attend some more comment to both `sendResponse` and also `dispatchRequest` function, could you please take a look if it's better?

tgodzik reviewed (changes requested):
Thanks for the contribution! This looks really good! I think we should indeed try to add attach to bsp. Let me know if you need help with the PR there.

marek1840 commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -586,6 +586,9 @@ final class BloopBspServices(
           convert[bsp.ScalaMainClass](main => DebuggeeRunner.forMainClass(projects, main, state))
         case bsp.DebugSessionParamsDataKind.ScalaTestSuites =>
           convert[List[String]](filters => DebuggeeRunner.forTestSuite(projects, filters, state))
+        // TODO incorporate into BSP, maybe?
```
Agreed, looks like a natural progression for the debug API

marek1840 reviewed (approved):
Looks good!

pvid commented:
> Thanks for the contribution! This looks really good! I think we should indeed try to add attach to bsp. Let me know if you need help with the PR there.

I opened a PR in https://github.com/build-server-protocol/build-server-protocol/pull/139 to add the needed constants there

jvican commented:
Thanks for the contribution @pvid! Let me know when this PR has caught up with the BSP release and I'll have a look at it.

tgodzik commented:
I think the changes should we available under the snapshot `2.0.0-M12+45-f04150db-SNAPSHOT`

pvid commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -280,7 +280,8 @@ final class BloopBspServices(
               jvmTestEnvironmentProvider = Some(true),
-              jvmRunEnvironmentProvider = Some(true)
+              jvmRunEnvironmentProvider = Some(true),
+              canReload = Some(false)
```
@jvican not sure about right thing to do here - `canReload = Some(false)` or `canReload = None`?

tgodzik replied:
I think `false` is correct here, since that issue was introduced for sbt and is not used in Bloop

pvid commented:
> Thanks for the contribution @pvid! Let me know when this PR has caught up with the BSP release and I'll have a look at it.

Hi! Sorry for the longer pause, I have been enjoying some time off. I rebased my branch onto `scalacenter:master` and updated the code to use the latest BSP release

tgodzik reviewed (approved):
LGTM! Thanks for the great contribution!

tgodzik commented:
@jvican I think it's good to merge, what do you think?
